In [ ]:
import numpy as np
import pandas as pd

Идея: приравниваем друг к другу теоритические и выборочные моменты и выразим нужный параметр.

Пусть $X = (x_1, ... , x_n)$ - выборка из неизвестного распределения $ P \in (P_\theta | \theta \in \Theta), \Theta \subset \mathbb {R}^d$. Составляем систему и решаем её. Решение системы относительно $\theta$ называется оценкой по методу моментов.

$\begin{equation*}
 \begin{cases}
   E_\theta X_1 = \overline x ,
   \\
   E_\theta X_1 = \overline x^2 ,
   \\
   ...
   \\
   E_\theta X^d_1 = \overline x^d ,
 \end{cases}
\end{equation*}$

Распишем оценку по методу моментов, рассмотрим как вектор:

$m(\theta) = \begin{pmatrix}
                E_\theta g( x_1 ),
                \\
                ...
                \\
                E_\theta g( x^d_1 ),
              \end{pmatrix} =
              \begin{pmatrix}
                \overline g(x_1) ,
                \\
                ...
                \\
                \overline g(x^d) ,
              \end{pmatrix} = \overline{g(x)}$ \
отсюда \
$\hat{\theta} = m^{-1} (\overline{g(x))}$

Свойства:
1. Если $m^{-1}$ непрерывна, то $\hat{\theta}$ сильно состоятельная оценка $\theta$.
2. Если $m^{-1}$ непрерывна дифференцируема и $ E_\theta g^2_i(x_1) < \infty $, то $\hat{\theta}$ ассимптотически нормальная оценка $\theta$.

### Задача №1. Распределение Пуассона.


Число семян сорняков в пробах зерна подчинено закону **Пуассона**. Имеется выборка
N =130 проб зерна. Результаты записаны в таблице Т1. Найти параметр λ по выборке
методом моментов.

In [ ]:
T1 = pd.DataFrame({'x_i': [0, 1, 2, 3, 4, 5], 'n_i': [9, 39, 40, 24, 11, 7]}).T
T1

0   1   2   3   4  5
x_i  0   1   2   3   4  5
n_i  9  39  40  24  11  7

**Решение:**

Как известно, параметр λ для закона Пуассона – это математическое ожидание,которое по методу моментов оценивается первым выборочным моментом:

$\lambda = \overline x_{выб} = \frac{1}{N} \sum\limits_{i=1}^n x_i n_i$


In [ ]:
N = 130
l = (T1.iloc[0, :].values * T1.iloc[1, :].values).sum() / N
l

2.076923076923077

**Ответ:** $\lambda$ = 2.077

### Задача №2. Равномерное распределение.

При условии **равномерного распределения** случайной величины Х
произведена выборка. Результаты записаны в таблице Т2. Найти оценку параметров a и b.

In [ ]:
T2 = pd.DataFrame({'x_i': [3, 5, 7, 9, 11, 13, 15, 17, 19, 21 ], 'n_i': [21, 16, 15, 26, 22, 14, 21, 22, 18, 25]}).T
T2

0   1   2   3   4   5   6   7   8   9
x_i   3   5   7   9  11  13  15  17  19  21
n_i  21  16  15  26  22  14  21  22  18  25

**Решение:**

Приравняем математическое ожидание $EX = \frac{a+b}{2}$ и дисперсию $DX = \frac{(b-a)^2}{12}$ соответствующим моментами и получим систему: 

$\begin{equation*}
 \begin{cases}
   EX = \frac{a+b}{2} = \mu_1 = \frac{1}{N} \sum\limits_{i=1}^n x_i n_i,
   \\
   DX = \frac{(b-a)^2}{12} = \mu_2 = \frac{1}{N} \sum\limits_{i=1}^n x^2_i n_i - \Big ( \frac{1}{N} \sum\limits_{i=1}^n x_i n_i \Big )^2 .
 \end{cases}
\end{equation*}$

$\mu_1 = \frac{1}{N} \sum\limits_{i=1}^n x_i n_i$

In [ ]:
N = T2.iloc[1, :].values.sum()
all_x = T2.iloc[0, :].values
all_n = T2.iloc[1, :].values
u_1 = (all_x * all_n).sum() / N
u_1

12.31

$ \mu_2 = \frac{1}{N} \sum\limits_{i=1}^n x^2_i n_i - \Big ( \frac{1}{N} \sum\limits_{i=1}^n x_i n_i \Big )^2 $

In [ ]:
u_2 = ((all_x ** 2 * all_n).sum() / N ) - u_1 ** 2
u_2

33.78389999999999

Подставляем в систему:

$\begin{equation*}
 \begin{cases}
   \frac{a+b}{2} = \mu_1 ,
   \\
   \frac{(b-a)^2}{12} = \mu_2 .
 \end{cases}
\end{equation*}$

$\begin{equation*}
 \begin{cases}
   \frac{a+b}{2} = 12.31 ,
   \\
   \frac{(b-a)^2}{12} = 33.78 .
 \end{cases}
\end{equation*}$

$\begin{equation*}
 \begin{cases}
   \frac{a+b}{2} = 12.31 ,
   \\
   (b-a)^2 = 12 х 33.78 = 405.36 .
 \end{cases}
\end{equation*}$

$\begin{equation*}
 \begin{cases}
   \frac{a+b}{2} = 12.31 ,
   \\
   b-a = \sqrt{405.36}.
 \end{cases}
\end{equation*}$

$\begin{equation*}
 \begin{cases}
   a = 2х12.31 - b = 24.62 - b,
   \\
   b-a = \sqrt{405.36}.
 \end{cases}
\end{equation*}$

$\begin{equation*}
 \begin{cases}
   a = 2х12.31 - b = 24.62 - b,
   \\
   b- 24.62 + b = \sqrt{405.36}.
 \end{cases}
\end{equation*}$

$2b - 24.62 = \sqrt{405.36} $ \

$ b = \frac{\sqrt{405.36} + 24.62}{2} \approx 22.38$

$a = 24.62 - 22.38 \approx 2.24 $

**Ответ:** \
$ a \approx 2.24 $ \
$ b \approx 22.38 $

### Задача №3. Биноминальное распределение.

Найти методом моментов по выборке $x_1, ..., x_n$ точечную оценку параметра $p$ биномиального распределения $P_m (x_i) = C_m^{x_i} p^{x_i}(1-p)^m-x_i$, где $x_i$ - число появлений события в $i$-ом опыте ($i = 1,2,.., n$), $m$ - количество испытаний в одном опыте. 

**Решение:**

Требуется оценить один параметр, поэтому достаточно иметь одно уравнение
относительно этого параметра. Математическое ожидание распределения Бернулли равно $EX = np$. Приравняем теоритический момент к выборочному среднему.

$EX = np = \overline x_{выб} = \frac{1}{m} \sum\limits_{i=1}^m x_i n_i$, отсюда

$p = \frac{1}{mn} \sum\limits_{i=1}^m x_i n_i = \frac{\overline x_{выб}}{m}$

**Ответ:** $ p = \frac{\overline x_{выб}}{m}$

### Задача №4. Нормальное распределение.

Найти методом моментов по выборке $x_1, ..., x_n$ точечные оценки неизвестных
параметров $a$ и $σ$ **нормального распределения**, плотность которого $f(x) = \frac{1}{\sigma \sqrt{2 \pi}} \exp^{-(x-a)^2 / (2\sigma)^2} $

**Решение:**

Для отыскания двух неизвестных параметров необходимо иметь два уравнения:
приравняем начальный теоретический момент первого порядка и центральный
теоретический момент второго порядка соответствующим эмпирическим моментам.

$ EX = a = \overline x_{выб} = \frac{1}{n} \sum\limits_{i=1}^n x_i n_i $ \

$ EX^2 = \sigma^2 = D_{выб} = \frac{1}{n} \sum\limits_{i=1}^n x^2_i n_i - \Big ( \frac{1}{n} \sum\limits_{i=1}^n x_i n_i \Big )^2 $, отсюда

**Ответ:**

$ a = \frac{1}{n} \sum\limits_{i=1}^n x_i n_i $ \

$ \sigma = \sqrt{D_{выб}} = \sqrt{ \frac{1}{n} (x_i - \overline x_{выб})^2 n_i}$

### Задача №5. Равномерное распределение.

Пусть случайная величина $\xi$ имеет плотность $p(x) = \frac{1}{b-a}$, если $x \in (a;b)$, и $p(x)=0$, если $x \notin (a;b)$. Произведена выборка T3. Используя метод моментов, найти a и b . 

In [ ]:
T3 = pd.DataFrame({'x_i': [1, 2, 3, 4, 5, 8, 9], 'n_i': [3, 5, 4, 3, 6, 4, 5]}).T
T3

0  1  2  3  4  5  6
x_i  1  2  3  4  5  8  9
n_i  3  5  4  3  6  4  5

**Решение:** 

$\begin{equation*}
 \begin{cases}
   EX = \frac{a+b}{2} = \overline x_{выб} = \frac{1}{N} \sum\limits_{i=1}^n x_i n_i,
   \\
   DX = \frac{(b-a)^2}{12} = D_{выб} = \frac{1}{N} \sum\limits_{i=1}^n x^2_i n_i - \Big ( \frac{1}{N} \sum\limits_{i=1}^n x_i n_i \Big )^2 .
 \end{cases}
\end{equation*}$

In [ ]:
N = T3.iloc[1, :].values.sum()
all_x = T3.iloc[0, :].values
all_n = T3.iloc[1, :].values
x_s = (all_x * all_n).sum() / N
x_s

4.8

In [ ]:
D_s = ((all_x ** 2 * all_n).sum() / N ) - x_s ** 2
D_s

7.560000000000002

Подставляем в систему:

p.s решение системы смотреть в задаче №2

$\begin{equation*}
 \begin{cases}
   \frac{a+b}{2} = 4.8 ,
   \\
   \frac{(b-a)^2}{12} = 7.56 .
 \end{cases}
\end{equation*}$

**Ответ:** \
$ a \approx 0.038 $ \
$ b \approx 9.562 $

### Задача №6. Нормальное распределение.

Из продукции станка извлечена выборка изделий и результаты измерений одного из размеров изделий (в мм) приведены ниже. Предполагая, что выборка значений контролируемого размера изделий подчиняется **нормальному закону распределения**, найдите оценку дисперсии с помощью метода моментов.

**Решение:**

$EX = \sigma^2 = D_{выб} = \frac{1}{n} \sum\limits_{i=1}^n x^2_i n_i - \Big ( \frac{1}{n} \sum\limits_{i=1}^n x_i n_i \Big )^2 $

In [ ]:
sample = np.array([17.4, 19.9, 19.9, 19.0, 20.7, 18.2, 18.6, 20.3, 21.7, 20.8, 16.7, 20.3, 22.5, 18.3, 21.1, 16.3, 22.0, 19.5, 18.1, 21.1, 20.0, 21.6, 21.5, 21.1, 23.3, 23.0, 17.8, 19.5, 21.6, 22.9])

In [ ]:
n = len(sample)

In [ ]:
D_s = ((sample**2).sum() / n ) - ( sample.sum() / n )**2
D_s

3.413788888888746

**Ответ:** $D_{выб}$ = 3.414

### Задача №7. Равномерное распределение.

По выборке $(x,y,z,t,k,s)$ с помощью метода моментов найдите точечную оценку параметра $\theta$ равномерно на $[0,\theta]$ распределенной случайной величины

$f_\xi(x) = \begin{equation*}
        \begin{cases}
          \frac{1}{\theta}, x \in [0, \theta],
          \\
          0 , x \notin [0, \theta]   .
        \end{cases}
        \end{equation*}$

Ответ записать в виде функции от $x,y,z,t,k,s$

**Решение:**

Кол-во элементов в выборке $(x,y,z,t,k,s)$: \
$n = 6$

Теоритическое мат. ожидания равномерного распределения равно: \
$EX = \frac{a + b}{2}$ \

В нашем случае: \
$EX = \frac{0 + \theta}{2}$ \
тогда, \
$EX = \frac{\theta}{2} = \overline x_{выб} \Rightarrow \theta = 2 \overline x_{выб} = 2 \frac{1}{n} \sum\limits_{i=1}^n x_i$

**Ответ:**

$\theta = 2 \frac{x+ y + z + t + k + s}{6}$



### Задача №8. Нормальное распределение.

По выборке $x,y,z,t,s$ найдите точечную оценку стандартного отклонения $\sigma$ нормально распределенной случайной величины, если мат.ожидание известно и равно $a$.

В качестве ответа укажите функция от $x, y, z, t, s, a$

**Решение:**

$EX = a$

$ EX^2 = \sigma^2 = D_{выб} = \frac{1}{N} \sum\limits_{i=1}^n x^2_i - \Big ( \frac{1}{N} \sum\limits_{i=1}^n x_i  \Big )^2 $, \
отсюда \

$\sigma = \sqrt{ \frac{1}{N} \sum\limits_{i=1}^n x^2_i  - \Big ( \frac{1}{N} \sum\limits_{i=1}^n x_i \Big )^2 } = \sqrt{ \frac{1}{N} \sum\limits_{i=1}^n x^2_i  - a^2 }  = \sqrt{ \frac{x^2 + y^2 + z^2 + t^2 + s^2}{5} - a^2 }$

**Ответ:** \
$ \sqrt{ \frac{x^2 + y^2 + z^2 + t^2 + s^2}{5} - a^2 } $